experiments/sandbox-next/src/utils/codegen/core/scanner.ts

```diff
@@ -0,0 +1,238 @@
+import { readdir } from "fs/promises";
+import { readFileSync } from "fs";
+import { join } from "path";
+import type {
+  ComponentInfo,
+  ScanResult,
+  ValidationResult,
+  ComponentValidation,
+  CodegenConfig,
+} from "../types";
+import { CodegenError } from "../utils/errors";
+
+interface ValidationCollector {
+  errors: string[];
+  warnings: string[];
+  componentValidation?: ComponentValidation;
+  transformerValidation?: ComponentValidation;
+}
+
+export class Scanner {
+  private static readonly INLINE_DEFAULT_EXPORT_PATTERN =
+    /export\s+default\s+[^;]+/;
+  private static readonly NAMED_DEFAULT_EXPORT_PATTERN =
+    /export\s*{\s*\w+\s+as\s+default\s*}/;
+
+  private readonly config: CodegenConfig;
+
+  constructor(config: CodegenConfig) {
+    this.config = config;
+  }
+
+  public async scan(): Promise<ScanResult> {
+    const result: ScanResult = {
+      components: [],
+      skippedFolders: [],
+      transformerWarnings: [],
+    };
+
+    try {
+      const entries = await readdir(this.config.componentsDir, {
+        withFileTypes: true,
+      });
+
+      for (const entry of entries) {
+        if (!entry.isDirectory()) continue;
+
+        const componentDir = join(this.config.componentsDir, entry.name);
+        const componentFiles = await readdir(componentDir);
+
+        const fileValidation = this.validateComponent(
+          componentDir,
+          componentFiles,
+          this.config
+        );
+        if (!fileValidation.isValid) {
+          result.skippedFolders.push(entry.name);
+          result.transformerWarnings.push(...fileValidation.errors);
+          continue;
+        }
+
+        const transformerFile = componentFiles.find((file) =>
+          file.includes("transformer.tsx")
```

Comment:
consider parametrize this, what happens is i wanna use another name file or another extension like .ts?